common/serial/serial.h

```diff
@@ -0,0 +1,219 @@
+#define COM4_IO_PORT 0x2e8
+
+/* Limits */
+#define MIN_DATA_SIZE 5
+#define MAX_DATA_SIZE 8
+
+#define MIN_STOP_BITS 2
+#define MAX_STOP_BITS 2
+
+#define MIN_INTERRUPT 0
+#define MAX_INTERRUPT 15
+
+/* Baud rate defines and lookup tables */
+#define MIN_BAUD_RATE 50
+#define MAX_BAUD_RATE 115200
+#define DEFAULT_BAUD_RATE 9600
+
+struct __divisor_rate {
+    uint32_t baud_rate;
+    uint16_t divisor;
+};
+
+extern struct __divisor_rate divisor_table[];
+
+/* Parity defines and lookup tables */
+typedef enum {
+    NONE,
+    ODD,
+    EVEN,
+    MARK,
+    SPACE,
+    __PARITY_MAX
+} PARITY_MODE;
+
+#define PARITY_MASK             0xC7
+
+struct __parity_bits {
+    PARITY_MODE mode;
+    uint8_t parity_bits;
+};
+
+extern struct __parity_bits parity_table[];
+
+/* Data size defines and lookup tables */
+#define DATA_SIZE_MASK          0xFC
+
+struct __data_bits {
+    uint8_t data_size;
+    uint8_t data_bits;
+};
+
+extern struct __data_bits data_size_table[];
+
+/* Receive buffer register */
+#define RBR_OFFSET              0x00
+/* Transmit hold register */
+#define THR_OFFSET              0x00
+ /* Interrupt Enable register */
+#define IER_OFFSET              0x01
+/* Divisor latch registers */
+#define DL_LSB_OFFSET           0x00
+#define DL_MSB_OFFSET           0x01
+/* Interrupt ID register */
+#define IID_OFFSET              0x02
+/* FIFO Control register */
+#define FCR_OFFSET              0x02
+#define FCR_MASK                (1<<0)
+
+/* Line control register */ 
+#define LCR_OFFSET              0x03
+/* Modem control register */    
+#define MCR_OFFSET              0x04
+
+/****************************************
+ *   Line status register               *
+ ****************************************/
+// IO Offset
+#define LSR_OFFSET              0x05
+// Data ready flag
+#define LSR_DR                  (1<<0)
+// Overrun error flag
+#define LSR_OE                  (1<<1)
+// Parity error flag
+#define LSR_PE                  (1<<2)
+// Frame error flag
+#define LSR_FE                  (1<<3)
+// Break indicator flag
+#define LSR_BI                  (1<<4)
+// Transmit hold register empty flag
+#define LSR_THRE                (1<<5)
+// Transmit empty flag
+#define LSR_TEMT                (1<<6)
+// Error byte in FIFO
+#define LSR_EBIF                (1<<7)
+
+
+/* Modem status register */ 
+#define MSR_OFFSET              0x06
+#define SCRATCH_REG_OFFSET      0x07
+
+/* Divisor latch access bit 
+    
+    This is set in the LCR to mux
+    the operational mode of the first
+    two registers.
+*/
+#define DLAB                    ((uint8_t)0x80)
+
+#define DATA_SIZE(x) (x-MIN_DATA_SIZE)
+
+class serial_port : public port_io {
+
+public:
+    serial_port(uint8_t port = 1, uint32_t baud = DEFAULT_BAUD_RATE, uint8_t data_bits = 8, 
+                PARITY_MODE parity = NONE, uint8_t stop_bits = 1);
+    ~serial_port(void);
+
+    int32_t open(void);
+    int32_t close(void);
+
+    void set_baud_rate(uint32_t baud);
```

Comment:
While I'm generally against setters doing any error reporting, I think it's reasonable in these scenarios (the value is expected to belong to a small subset of the types domain).